engine/storage/image/src/org/apache/cloudstack/storage/image/store/TemplateObject.java

```diff
@@ -189,12 +189,24 @@ public void processEvent(ObjectInDataStoreStateMachine.Event event, Answer answe
                     TemplateObjectTO newTemplate = (TemplateObjectTO)cpyAnswer.getNewData();
                     VMTemplateStoragePoolVO templatePoolRef = templatePoolDao.findByPoolTemplate(getDataStore().getId(), getId());
                     templatePoolRef.setDownloadPercent(100);
-                    if (newTemplate.getSize() != null) {
+
+                    // In the case of managed storage, the template size may already be specified (by the storage plug-in), so do not overwrite it.
+                    if (templatePoolRef.getTemplateSize() == 0 && newTemplate.getSize() != null) {
                         templatePoolRef.setTemplateSize(newTemplate.getSize());
                     }
+
                     templatePoolRef.setDownloadState(Status.DOWNLOADED);
-                    templatePoolRef.setLocalDownloadPath(newTemplate.getPath());
-                    templatePoolRef.setInstallPath(newTemplate.getPath());
+
+                    // In the case of managed storage, the local download path may already be specified (by the storage plug-in), so do not overwrite it.
+                    if (templatePoolRef.getLocalDownloadPath() == null) {
+                        templatePoolRef.setLocalDownloadPath(newTemplate.getPath());
+                    }
+
+                    // In the case of managed storage, the install path may already be specified (by the storage plug-in), so do not overwrite it.
+                    if (templatePoolRef.getInstallPath() == null) {
```

Comment:
Consider using ``Strings.isNullorEmpty`` to cover both the ``null`` and empty cases.